RFC.md

```diff
@@ -1,12 +1,234 @@
+multiple dimensions most languages take, which we discuss to motivate 
+the need for a multi-dimensional array index. This consideration impacts
+the kinds of analysis we can perform on the program. In Polly, we care about
+dependence analysis, so the examples here will focus on that particular problem.
 
-# The math, how to express every conceivable array indexing scheme known to man
+Let us consider an array indexing operation of the form:
+```cpp
+int ex1(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+	assert(x1 != x2);
+	assert(y1 != y2);
+	B[x1][y1] = 1;
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
 
+One would like to infer that since the array indeces _interpreted as tuples_
+`(x1, y1)` and `(x2, y2)` do not alias, due to the guarding asserts
+that `x1 != x2` and `y1 != y2`. the write `B[x1][y1] = 1` can 
+in no way interfere with the value of `B[x2][y2]`. Consquently,
+we can optimise the program into:
 
+
+```cpp
+int ex1_opt(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
+
+However, alas, this is illegal, for the C language does not provide these
+semantics. It is conceivable that `x1 != x2, y1 != y2`, but the indeces
+do actually alias, since according to C semantics, the two indeces alias
+if the _flattened representation of the indeces alias_. So, in this case,
+pick the parameter values:
+
+```
+n = m = 3
+x1 = 1, y1 = 0; B[x1][y1] = nx1+y1 = 3*1+0=3
+x2 = 0, y2 = 3; B[x2][y2] = nx2+y2 = 3*0+3=3
+```
+
+Hence, the arrays _do alias_, and the transformation proposed in `ex1_opt`
+is unsound in general.
+
+However, there are languages that provide this "tuple-indexed" semantics,
+where one can infer that the indexing:
+`[x1][y1] != [x2][y2]` iff `x1 != x2 || y1 != y2`.
+
+Julia, Fortran, and Chapel are examples of such languages which target
+LLVM.
+
+Currently, the LLVM semantics of `getelementptr` talk purely of
+the C style flattened views of arrays. This inhibits the ability of the optimiser
+to understand the multidimensional examples as given above, and we
+are forced to make conservative assumptions, inhibiting optimisation. 
+This information must ideally be expressible in LLVM, such that LLVM's optimisers and
+alias analysis can use this information to model multidimensional-array
+semantics.
+
+
+
+## Evaluation of the impact of the intrinsic on accuracy of dependence analysis
+
+- This has been implemented in an exprimental branch of Polly, and was used 
+on the COSMO climate weather model. This greatly helped increase the accuracy
+of Polly's analysis, since we eliminated the guessing game from the array analysis.
+
+- This has also been implemented as part of a GSoC effort to unify
+Chapel and Polly. 
+
+- Molly, the distributed version of Polly written by Michael Kruse for his PhD
+  also implemented a similar scheme. In his use-case, guessing as we perform
+  optimistically was not possible, so this kind of intrinsic was _necessary_
```

Comment:
"guessing as we perform optimistically was not possible" - there seems to be something wrong with this sentence. Maybe "In his use-case, optimistic delinearization with runtime checks was not possible, so this..."